AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m

```diff
@@ -68,22 +68,41 @@ - (BOOL)saveLog:(id<MSLog>)log withGroupId:(NSString *)groupId flags:(MSFlags)fl
   }
   return [self executeQueryUsingBlock:^int(void *db) {
            int result = [MSDBStorage executeNonSelectionQuery:addLogQuery inOpenedDatabase:db];
+           NSMutableArray<NSNumber *> *logsCanBeDeleted = nil;
+           if (result == SQLITE_FULL) {
+
+             // Selecting logs with equal or lower priority and ordering by priority then age.
+             NSString *query = [NSString stringWithFormat:@"SELECT \"%@\" FROM \"%@\" WHERE \"%@\" <= %u ORDER BY \"%@\" ASC, \"%@\" ASC",
+                                                          kMSIdColumnName, kMSLogTableName, kMSPriorityColumnName, (unsigned int)flags,
+                                                          kMSPriorityColumnName, kMSIdColumnName];
+             NSArray<NSArray *> *entries = [MSDBStorage executeSelectionQuery:query inOpenedDatabase:db];
+             logsCanBeDeleted = [NSMutableArray new];
+             for (NSMutableArray *row in entries) {
+               [logsCanBeDeleted addObject:row[0]];
+             }
+           }
 
            // If the database is full, delete logs until there is room to add the log.
            long countOfLogsDeleted = 0;
-           while (result == SQLITE_FULL) {
-             result = [MSLogDBStorage deleteOldestLogsWithCount:1 inOpenedDatabase:db];
+           NSUInteger index = 0;
+           while (result == SQLITE_FULL && index < [logsCanBeDeleted count]) {
+             result = [MSLogDBStorage deleteLogsFromDBWithColumnValues:@[ logsCanBeDeleted[index++] ]
+                                                            columnName:kMSIdColumnName
+                                                      inOpenedDatabase:db];
              if (result != SQLITE_OK) {
                break;
              }
              ++countOfLogsDeleted;
              result = [MSDBStorage executeNonSelectionQuery:addLogQuery inOpenedDatabase:db];
            }
            if (countOfLogsDeleted > 0) {
-             MSLogDebug([MSAppCenter logTag], @"Log storage was over capacity, %ld oldest log(s) deleted.", (long)countOfLogsDeleted);
+             MSLogDebug([MSAppCenter logTag], @"Log storage was over capacity, %ld oldest log(s) with equal or lower priority deleted.",
+                        (long)countOfLogsDeleted);
            }
            if (result == SQLITE_OK) {
              MSLogVerbose([MSAppCenter logTag], @"Log is stored with id: '%ld'", (long)sqlite3_last_insert_rowid(db));
+           } else if (result == SQLITE_FULL && index == [logsCanBeDeleted count]) {
+             MSLogDebug([MSAppCenter logTag], @"No logs with equal or lower priority found and storage already full; discarding the log.");
```

Comment:
wording - "No logs with equal or lower priority **were** found and storage **is** already full"